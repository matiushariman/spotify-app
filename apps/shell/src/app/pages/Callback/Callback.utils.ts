import { AuthUtils } from '@spotify-app/auth-utils';

export const setTokenFromUrlParams = async (hash: string) => {
  const authUtils = new AuthUtils();
  const splitHash = hash.split('access_token=');
  const accessToken = splitHash[1].split('&');

  authUtils.setToken({
    token: accessToken[0],
    privateKey: process.env['NX_PRIVATE_KEY'] ?? '',
  });
};
