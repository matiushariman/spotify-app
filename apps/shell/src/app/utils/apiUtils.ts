import type { ErrorResponse } from '@spotify-app/api-client';
import { AuthUtils } from '@spotify-app/auth-utils';

export const handleApiError = (err: ErrorResponse) => {
  const {
    error: { status },
  } = err;

  if (status === 401) {
    window.alert(
      'Your session has expired! Click close to reauthenticate your session.'
    );
    const authUtils = new AuthUtils();
    authUtils.removeToken();

    window.location.replace('/');
  }
};
