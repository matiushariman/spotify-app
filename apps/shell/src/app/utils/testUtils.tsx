import { FC, ReactElement, ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import {
  RouterProvider,
  createMemoryRouter,
  RouteObject,
} from 'react-router-dom';
import { QueryClientProvider, QueryClient } from 'react-query';

import '../../styles.css';

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  const routes: RouteObject[] = [
    {
      path: '/',
      element: children,
    },
  ];
  const router = createMemoryRouter(routes);

  return (
    <QueryClientProvider client={new QueryClient()}>
      <RouterProvider router={router} />
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
