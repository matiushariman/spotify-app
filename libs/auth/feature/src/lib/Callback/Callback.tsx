import { Button } from '@react-spotify/shared-ui-components';
import { useSessionStore } from '@react-spotify/shared-stores';
import { useLocation, useNavigate } from 'react-router-dom';
import { useEffect } from 'react';

export function Callback() {
  const { hash } = useLocation();
  const navigate = useNavigate();
  const setAccessToken = useSessionStore((state) => state.setAccessToken);

  useEffect(() => {
    if (hash) {
      const accessToken = hash.split('&')[0].split('=')[1];
      setAccessToken(accessToken);
      navigate('/');
    }
  }, []);

  return (
    <div className="flex h-screen items-center justify-around p-4">
      {!hash ? (
        <div className="space-y-6">
          <div className="space-y-2">
            <h1 className="text-center">Token not found!</h1>
            <p className="text-gray-400">
              Click the button below to go back to login page.
            </p>
          </div>
          <div className="text-center">
            <Button onClick={() => navigate('/', { replace: true })}>
              Back to Login
            </Button>
          </div>
        </div>
      ) : (
        <h1>Redirecting you back to the app...</h1>
      )}
    </div>
  );
}

export default Callback;
