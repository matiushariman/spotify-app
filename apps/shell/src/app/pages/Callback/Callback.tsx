import { useEffect } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

import { setTokenFromUrlParams } from './Callback.utils';

export const Callback = () => {
  const { hash } = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    if (hash) {
      setTokenFromUrlParams(hash).then(() => {
        navigate('/');
      });
    }
  }, [hash]);

  return <p>Redirecting you back to the app..</p>;
};
