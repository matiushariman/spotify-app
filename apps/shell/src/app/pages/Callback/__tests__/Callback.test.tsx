import { Callback } from '../Callback';
import { render, waitFor } from '../../../utils/testUtils';

const mockUseNavigate = jest.fn();

jest.mock('./Callback.utils', () => ({
  setTokenFromUrlParams: async () => jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockUseNavigate,
  useLocation: jest
    .fn()
    .mockReturnValueOnce({
      hash: null,
    })
    .mockReturnValueOnce({
      hash: '#access_token=abcd&',
    }),
}));

describe('pages/Callback', () => {
  it('should render Callback that matches snapshot if hash is empty', () => {
    const { baseElement } = render(<Callback />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should call navigate after finish setting token', async () => {
    render(<Callback />);

    await waitFor(() => expect(mockUseNavigate).toHaveBeenCalled());
  });
});
