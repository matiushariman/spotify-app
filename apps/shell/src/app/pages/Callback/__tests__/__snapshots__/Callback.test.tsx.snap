// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Callback should render Callback that matches snapshot 1`] = `
<body>
  <div>
    <p>
      Redirecting you back to the app..
    </p>
  </div>
</body>
`;

exports[`pages/Callback should render Callback that matches snapshot if hash is empty 1`] = `
<body>
  <div>
    <p>
      Redirecting you back to the app..
    </p>
  </div>
</body>
`;
